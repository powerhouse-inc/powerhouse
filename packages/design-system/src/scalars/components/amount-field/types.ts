export type Amount = number | undefined;
export type AmountPercentage = number | undefined;
export interface AmountCurrency {
  amount?: number;
  currency: string;
}
export interface AmountToken {
  amount?: bigint;
  token: string;
}

export type AmountFieldPropsGeneric =
  | {
      type: "Amount";
      value?: Amount;
      trailingZeros?: boolean;
    }
  | {
      type: "AmountCurrency";
      value?: AmountCurrency;
      trailingZeros?: boolean;
    }
  | {
      type: "AmountPercentage";
      value?: AmountPercentage;
      trailingZeros?: boolean;
    }
  | {
      type: "AmountToken";
      value?: AmountToken;
    };

export type AmountValue =
  | Amount
  | AmountPercentage
  | AmountCurrency
  | AmountToken;
