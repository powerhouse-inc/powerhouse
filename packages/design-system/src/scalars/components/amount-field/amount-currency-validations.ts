import { type AmountCurrency } from "../../../ui/components/data-entry/amount-input/types.js";
import { type AmountFieldProps } from "./amount-field.js";

export const validateAmountCurrency =
  ({ type, units }: AmountFieldProps) =>
  (value: unknown) => {
    if (!value) return true;
    if (typeof value === "object" && "unit" in value) {
      if (
        ["AmountFiat", "AmountCrypto", "AmountCurrency", "Amount"].includes(
          type,
        )
      ) {
        if (
          !units?.some(
            (currency) => currency.ticker === (value as AmountCurrency).unit,
          )
        ) {
          return "Please select a valid currency";
        }
      }
    }

    return true;
  };
