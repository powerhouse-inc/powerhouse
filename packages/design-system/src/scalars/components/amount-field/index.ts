export { AmountField, type AmountFieldProps } from "./amount-field.js";
