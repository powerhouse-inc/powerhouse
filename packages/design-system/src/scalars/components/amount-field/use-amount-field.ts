import { getLabelValueCurrenct } from "@/scalars/utils/utils";
import { AmountCurrency, AmountFieldPropsGeneric, AmountValue } from "./types";

interface UseAmountFieldProps {
  value?: AmountValue;
  defaultValue?: AmountValue;
  type: AmountFieldPropsGeneric["type"];
  allowedCurrencies?: string[];
}

export const useAmountField = ({
  value,
  defaultValue = 0,
  type,
  allowedCurrencies = [],
}: UseAmountFieldProps) => {
  const currentValue = value ?? defaultValue;

  const valueInput =
    type === "Amount" || type === "AmountPercentage"
      ? (currentValue as number)
      : (currentValue as AmountCurrency).amount;
  const isPercent = type === "AmountPercentage";
  const isCurrency = type === "AmountCurrency";

  const options = getLabelValueCurrenct(allowedCurrencies);
  const currency =
    type === "AmountCurrency" ? (currentValue as AmountCurrency).currency : "";

  const isSearchable = options.length >= 5;

  return {
    isPercent,
    isCurrency,
    isSearchable,
    options,
    valueInput,
    currency,
  };
};
