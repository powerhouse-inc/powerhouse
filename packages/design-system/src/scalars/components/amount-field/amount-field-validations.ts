import { type ValidatorResult } from "#scalars";
import {
  type Amount,
  type AmountCrypto,
  type AmountCurrency,
  type AmountFiat,
  type AmountInputPropsGeneric,
  type AmountValue,
} from "../../../ui/components/data-entry/amount-input/types.js";
import { isValidBigInt } from "../../../ui/components/data-entry/amount-input/utils.js";
import { isValidNumber } from "../number-field/number-field-validations.js";
import { type AmountFieldProps } from "./amount-field.js";

const isAmountCurrencyFiat = (
  type: AmountInputPropsGeneric["type"],
): type is "AmountFiat" => type === "AmountFiat";

const isAmountCurrencyCrypto = (
  type: AmountInputPropsGeneric["type"],
): type is "AmountCrypto" => type === "AmountCrypto";

const isAmountCurrencyUniversal = (
  type: AmountInputPropsGeneric["type"],
): type is "AmountCurrency" => type === "AmountCurrency";

const isAmount = (type: AmountInputPropsGeneric["type"]): type is "Amount" =>
  type === "Amount";

const getAmount = (
  value: AmountValue,
  type: AmountInputPropsGeneric["type"],
): number | bigint | undefined => {
  if (
    isAmountCurrencyFiat(type) ||
    isAmountCurrencyCrypto(type) ||
    isAmountCurrencyUniversal(type) ||
    isAmount(type)
  ) {
    if (!value) return undefined;
    return (
      (value as AmountFiat | AmountCrypto | AmountCurrency | Amount).amount ??
      undefined
    );
  }
  return value as number;
};

export const validateAmount =
  ({
    type,
    required,
    minValue,
    maxValue,
    allowNegative,
    units,
  }: AmountFieldProps) =>
  (value: unknown): ValidatorResult => {
    const amount = getAmount(
      value as AmountValue,
      type as AmountInputPropsGeneric["type"],
    );
    if (value === "") return true;
    if (amount?.toString() === "") {
      if (required) {
        return "This field is required";
      }
      return true;
    }

    if (amount === undefined) {
      if (required) {
        return "This field is required";
      }
      return true;
    }
    if (!isValidNumber(amount) && type !== "AmountCurrency") {
      return "Value is not a valid number";
    }
    if (!allowNegative && amount < 0) {
      return "Value must be positive";
    }
    if (type === "AmountCrypto") {
      if (!isValidBigInt(amount.toString())) {
        return "Value is not an bigint";
      }
      return true;
    }
    if (type === "AmountCurrency") {
      if (!isValidNumber(amount)) {
        return "Value is not a valid number";
      }
      if (Math.abs(Number(amount)) > Number.MAX_SAFE_INTEGER) {
        const amountStr = amount.toString();
        if (!/^\d+$/.test(amountStr)) {
          return "Value is not a valid bigint";
        }
        return true;
      }
    }

    if (maxValue) {
      if (amount > maxValue) {
        return `This field must be less than ${maxValue}`;
      }
    }
    if (minValue) {
      if (amount < minValue) {
        return `This field must be more than ${minValue}`;
      }
    }
    if (
      Math.abs(Number(amount)) > Number.MAX_SAFE_INTEGER &&
      (type === "AmountFiat" ||
        type === "AmountPercentage" ||
        type === "Amount")
    ) {
      return "Value is too large for number";
    }

    return true;
  };
