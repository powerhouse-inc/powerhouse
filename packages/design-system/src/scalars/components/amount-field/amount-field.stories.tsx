import { commonCryptoCurrencies, commonFiatCurrencies } from "#scalars";
import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  PrebuiltArgTypes,
  StorybookControlCategory,
  getDefaultArgTypes,
  getValidationArgTypes,
} from "../../lib/storybook-arg-types.js";
import { AmountField } from "./amount-field.js";
const mappedFiatCurrencies = commonFiatCurrencies.map((currency) => ({
  ...currency,
  label: currency.ticker,
}));
const mappedCryptoCurrencies = commonCryptoCurrencies.map((currency) => ({
  ...currency,
  label: currency.ticker,
}));

const meta = {
  title: "Document Engineering/Scalars/Amount Field",
  component: AmountField,
  decorators: [withForm],
  parameters: {
    layout: "centered",
    form: {
      defaultValues: {
        "amount-field": {
          amount: undefined,
          unit: "",
        },
      },
    },
  },
  tags: ["autodocs"],
  argTypes: {
    units: {
      control: "object",
      description: "Array of Currency objects",
      table: {
        type: { summary: "Currency[]" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: "type",
        neq: "AmountPercentage",
      },
    },
    includeCurrencySymbols: {
      control: "boolean",
      description: " Whether to display currency symbols alongside codes",
      table: {
        defaultValue: { summary: "true" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: "type",
        neq: "AmountPercentage",
      },
    },
    symbolPosition: {
      control: "select",
      description: "Position of the currency symbol",
      options: ["left", "right"],
      table: {
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: "type",
        neq: "AmountPercentage",
      },
    },
    step: {
      control: "number",
      description: "The step value for the amount field",
      table: {
        defaultValue: { summary: "1" },
        type: { summary: "number" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    currencyPosition: {
      control: "select",
      options: ["left", "right"],
      description:
        "Determines the position of the currency select dropdown relative to the amount input field.",
      table: {
        defaultValue: { summary: "right" },
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: "type",
        neq: "AmountPercentage",
      },
    },

    trailingZeros: {
      control: "boolean",
      description:
        "When precision is set, for example to 2, determines if the the trailing zeros should be preserved ( for example: 25.00,7.50, etc.) or not ( for example: 25, 7.5).",
      if: {
        arg: "type",
        neq: "AmountToken",
      },
      table: {
        type: { summary: "boolean" },
        category: StorybookControlCategory.VALIDATION,
      },
    },
    viewPrecision: {
      control: "number",
      description: "Number of decimal places viewed",
      table: {
        type: { summary: "number" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    precision: {
      control: "number",
      description: "Number of decimal places viewed",
      table: {
        type: { summary: "number" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    allowNegative: {
      control: "boolean",
      description: "Whether negative values are allowed (true or false).",
      table: {
        defaultValue: { summary: "false" },
        type: { summary: "boolean" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    numberProps: {
      control: "object",
      description: "All the props options for number field",
      table: {
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    selectProps: {
      control: "object",
      description: "All the props options for select field",
      table: {
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    // TODO:Improve the AmountType descriptions for the value
    ...getDefaultArgTypes(),

    value: {
      control: "object",
      description:
        "The value of the amount field. Can be a number, an object with currency, or undefined. Examples: { amount: 100, currency: 'USD' }, 200, undefined.",
      table: {
        type: { summary: "object | number | undefined" },
        category: StorybookControlCategory.DEFAULT,
      },
    },
    ...PrebuiltArgTypes.placeholder,
    ...getValidationArgTypes(),
    ...PrebuiltArgTypes.minValue,
    ...PrebuiltArgTypes.maxValue,
    type: {
      control: "select",
      options: [
        "Amount",
        "AmountFiat",
        "AmountPercentage",
        "AmountCrypto",
        "AmountCurrency",
      ],
      description: "The type of amount field.",
      table: {
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
  },

  args: {
    name: "amount-field",
  },
} satisfies Meta<typeof AmountField>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    placeholder: "0",
    units: mappedFiatCurrencies,
    label: "Enter Amount and Select Currency",
    placeholderSelect: "CUR",
    type: "Amount",
    value: {
      amount: undefined,
      unit: "",
    },
  },
};

export const WithValue: Story = {
  args: {
    placeholder: "Enter Amount",
    placeholderSelect: "CUR",
    label: "Enter Amount",
    type: "Amount",
    value: {
      amount: 100,
    },
  },
};
export const WithAmount: Story = {
  parameters: {
    form: {
      defaultValues: {
        "amount-field": "",
      },
    },
  },
  args: {
    units: mappedFiatCurrencies,
    placeholder: "Enter Amount",
    placeholderSelect: "CUR",
    label: "Enter Amout ",
    type: "AmountFiat",
    value: {
      amount: 345,
      unit: "EUR",
    },
  },
};
export const CurrencyIcon: Story = {
  args: {
    units: mappedCryptoCurrencies,
    placeholder: "Enter Amount",
    label: "Enter Amount and Select Currency",
    type: "AmountCrypto",
    placeholderSelect: "CUR",
    value: {
      amount: 3454564564 as unknown as bigint,
      unit: "BTC",
    },
  },
};

export const WithToken: Story = {
  parameters: {
    units: mappedCryptoCurrencies,
    form: {
      defaultValues: {
        "amount-field": {
          amount: "",
          unit: "",
        },
      },
    },
  },
  args: {
    placeholder: "Enter Amount",
    label: "Enter Amount and Select Currency",
    type: "AmountCurrency",
    placeholderSelect: "CUR",
    units: [...mappedCryptoCurrencies, ...mappedFiatCurrencies],
    value: {
      amount: 123 as unknown as bigint,
      unit: "BTC",
    },
  },
};

export const WithValuePercent: Story = {
  parameters: {
    form: {
      defaultValues: {
        "amount-field": "",
      },
    },
  },
  args: {
    label: "Enter Percentage ",
    placeholder: "Enter Amount",
    type: "AmountPercentage",
    value: 9,
  },
};
export const Disable: Story = {
  args: {
    units: mappedCryptoCurrencies,
    label: "Enter Amount ",
    placeholder: "Enter Amount",
    type: "AmountFiat",
    placeholderSelect: "CUR",
    disabled: true,
    value: {
      amount: 9,
      unit: "ETH",
    },
  },
};

export const WithValueUniversalAmountCurrency: Story = {
  parameters: {
    form: {
      defaultValues: {
        "amount-field": {
          amount: 123,
          unit: "BTC",
        },
      },
    },
  },
  args: {
    units: [...mappedCryptoCurrencies, ...mappedFiatCurrencies],
    label: "Label",
    placeholder: "Enter Amount",
    placeholderSelect: "CUR",
    type: "AmountCurrency",
    value: {
      amount: 123,
      unit: "BTC",
    },
  },
};
