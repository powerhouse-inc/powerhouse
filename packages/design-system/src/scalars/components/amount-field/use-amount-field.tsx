import { TokenIcons } from "./amount-field";
import {
  AmountCurrencyCrypto,
  AmountCurrencyFiat,
  AmountCurrencyUniversal,
  AmountFieldPropsGeneric,
  AmountValue,
} from "./types";
import {
  displayValueAmount,
  getCountryCurrencies,
  getTokens,
  isNotSafeValue,
  isValidBigInt,
  isValidNumberGreaterThanMaxSafeInteger,
} from "./utils";
import { isValidNumber } from "../number-field/number-field-validations";
import { useMemo, useState } from "react";

interface UseAmountFieldProps {
  value?: AmountValue;
  defaultValue?: AmountValue;
  type: AmountFieldPropsGeneric["type"];
  allowedCurrencies?: string[];
  allowedTokens?: string[];
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void;
  tokenIcons?: TokenIcons;

  precision?: number;
  viewPrecision?: number;
  trailingZeros?: boolean;
}

export const useAmountField = ({
  value,
  defaultValue,
  type,
  allowedCurrencies = [],
  allowedTokens = [],
  onChange,
  onBlur,
  tokenIcons,
  precision,
  viewPrecision,
  trailingZeros,
}: UseAmountFieldProps) => {
  const currentValue = value ?? defaultValue;

  const baseValue = useMemo(() => {
    return currentValue === undefined
      ? undefined
      : type === "Amount" || type === "AmountPercentage"
        ? (currentValue as number | undefined)
        : type === "AmountCurrencyUniversal"
          ? (currentValue as AmountCurrencyUniversal).amount
          : (currentValue as AmountCurrencyFiat | AmountCurrencyCrypto).amount;
  }, [currentValue, type]);

  const valueSelect =
    currentValue === undefined
      ? undefined
      : type === "AmountCurrencyFiat"
        ? (currentValue as AmountCurrencyFiat).currency
        : type === "AmountCurrencyCrypto"
          ? (currentValue as AmountCurrencyCrypto).currency
          : type === "AmountCurrencyUniversal"
            ? (currentValue as AmountCurrencyUniversal).currency
            : undefined;

  const isBigInt =
    type === "AmountCurrencyCrypto" ||
    (type === "AmountCurrencyUniversal" &&
      isValidNumberGreaterThanMaxSafeInteger(baseValue?.toString()));

  const [inputFocused, setInputFocused] = useState(false);

  const rawAmountState = baseValue?.toString() ?? "";
  const displayValueAmountState = useMemo(() => {
    if (inputFocused) {
      return rawAmountState;
    }

    if (
      type === "Amount" ||
      type === "AmountPercentage" ||
      type === "AmountCurrencyFiat"
    ) {
      if (!isValidNumber(rawAmountState) || isNotSafeValue(rawAmountState)) {
        // Return the value without formatting if not valid
        return baseValue?.toString() ?? "";
      }
    }
    if (type === "AmountCurrencyCrypto") {
      return baseValue?.toString();
    }
    if (type === "AmountCurrencyUniversal") {
      if (isValidNumberGreaterThanMaxSafeInteger(rawAmountState)) {
        //   // Return the value without formatting if not valid
        return baseValue?.toString() ?? "";
      }
      if (!isValidNumber(rawAmountState) || isNotSafeValue(rawAmountState)) {
        // Return the value without formatting if not valid
        return baseValue?.toString() ?? "";
      }
    }
    // Here its safe to format the value
    return displayValueAmount(
      baseValue?.toString() ?? "",
      precision,
      viewPrecision,
      trailingZeros,
    );
  }, [
    inputFocused,
    type,
    baseValue,
    precision,
    viewPrecision,
    trailingZeros,
    rawAmountState,
  ]);

  const isPercent = type === "AmountPercentage";
  const isAmount = type === "Amount";

  //Allow select only if type is AmountCurrencyFiat or AmountCurrencyCrypto
  const isShowSelect =
    type === "AmountCurrencyFiat" ||
    type === "AmountCurrencyCrypto" ||
    type === "AmountCurrencyUniversal";

  // Filter only if allowedCurrencies is provided
  const optionsCurrencies = getCountryCurrencies(allowedCurrencies);
  const optionsTokenIcons = getTokens(allowedTokens, tokenIcons);

  // Options for AmountCurrencyUniversal type
  const optionsForUniversal =
    optionsCurrencies.length > 0
      ? optionsCurrencies
      : optionsTokenIcons.length > 0
        ? optionsTokenIcons
        : [];

  const options =
    type === "AmountCurrencyFiat"
      ? optionsCurrencies
      : type === "AmountCurrencyCrypto"
        ? optionsTokenIcons
        : // this can also currency fiat but its not possible to select the currency for simbol of the currency
          type === "AmountCurrencyUniversal"
          ? optionsForUniversal
          : [];
  // Handle the change of the input
  const handleOnChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.target.value;

    if (type === "AmountCurrencyFiat" && typeof value === "object") {
      const newValue = {
        ...value,
        amount: inputValue === "" ? undefined : inputValue,
      } as AmountValue;

      //Create the event
      const nativeEvent = new Event("change", {
        bubbles: true,
        cancelable: true,
      });
      Object.defineProperty(nativeEvent, "target", {
        value: { value: newValue },
        writable: false,
      });
      onChange?.(nativeEvent as unknown as React.ChangeEvent<HTMLInputElement>);
    }

    // Handle the change of the currency crypto
    if (type === "AmountCurrencyCrypto" && typeof value === "object") {
      const newValueCurrencyCrypto = {
        ...value,
        amount: inputValue === "" ? undefined : inputValue,
      } as AmountValue;

      //Create the event
      const nativeEvent = new Event("change", {
        bubbles: true,
        cancelable: true,
      });
      Object.defineProperty(nativeEvent, "target", {
        value: { value: newValueCurrencyCrypto },
        writable: false,
      });
      onChange?.(nativeEvent as unknown as React.ChangeEvent<HTMLInputElement>);
    }

    if (type === "Amount" || type === "AmountPercentage") {
      const amountValue = inputValue === "" ? undefined : inputValue;
      // Create the event
      const nativeEvent = new Event("change", {
        bubbles: true,
        cancelable: true,
      });
      Object.defineProperty(nativeEvent, "target", {
        value: { value: amountValue },
        writable: false,
      });
      onChange?.(nativeEvent as unknown as React.ChangeEvent<HTMLInputElement>);
    }
    // Handle the change of the universal amount
    if (type === "AmountCurrencyUniversal" && typeof value === "object") {
      const newValue = {
        ...value,
        amount: inputValue === "" ? undefined : inputValue,
      } as AmountValue;

      // Create the event
      const nativeEvent = new Event("change", {
        bubbles: true,
        cancelable: true,
      });
      Object.defineProperty(nativeEvent, "target", {
        value: { value: newValue },
        writable: false,
      });
      onChange?.(nativeEvent as unknown as React.ChangeEvent<HTMLInputElement>);
    }
  };
  // Handle the change of the select
  const handleOnChangeSelect = (e: string | string[]) => {
    let newValue:
      | AmountCurrencyFiat
      | AmountCurrencyCrypto
      | AmountCurrencyUniversal = {} as
      | AmountCurrencyFiat
      | AmountCurrencyCrypto;
    if (type === "AmountCurrencyFiat" && typeof value === "object") {
      newValue = {
        ...value,
        currency: typeof e === "string" ? e : undefined,
      } as AmountCurrencyFiat;
    }
    if (type === "AmountCurrencyCrypto" && typeof value === "object") {
      newValue = {
        ...value,
        currency: typeof e === "string" ? e : undefined,
      } as AmountCurrencyCrypto;
    }
    if (type === "AmountCurrencyUniversal" && typeof value === "object") {
      newValue = {
        ...value,
        currency: typeof e === "string" ? e : undefined,
      } as AmountCurrencyUniversal;
    }

    //Create the event
    const nativeEvent = new Event("change", {
      bubbles: true,
      cancelable: true,
    });
    Object.defineProperty(nativeEvent, "target", {
      value: { value: newValue },
      writable: false,
    });
    onChange?.(nativeEvent as unknown as React.ChangeEvent<HTMLInputElement>);
  };

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    setInputFocused(false);
    const inputValue = e.target.value;
    //Avoid transformation if the value is not a number
    if (type === "AmountCurrencyFiat" && typeof value === "object") {
      if (!isValidNumber(inputValue)) {
        const newValue = {
          ...value,
          amount: inputValue,
        };
        const nativeEvent = new Event("onBlur", {
          bubbles: true,
          cancelable: true,
        });
        Object.defineProperty(nativeEvent, "target", {
          value: { value: newValue },
          writable: false,
        });

        onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }

      // Avoid transformation if the value is not a number
      if (isNotSafeValue(inputValue)) {
        const newValue = {
          ...value,
          amount: inputValue,
        };
        onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }
      const formatValue = displayValueAmount(
        inputValue,
        precision,
        viewPrecision,
        trailingZeros,
      );
      // Update the state with the formatted value
      const newValue = {
        ...value,
        amount: formatValue,
      };
      onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
    }

    if (type === "AmountCurrencyCrypto" && typeof value === "object") {
      if (!isValidBigInt(inputValue) || isNotSafeValue(inputValue)) {
        const newValue = {
          ...value,
          amount: inputValue,
        };
        const nativeEvent = new Event("onBlur", {
          bubbles: true,
          cancelable: true,
        });
        Object.defineProperty(nativeEvent, "target", {
          value: { value: newValue },
          writable: false,
        });
        onBlur?.(nativeEvent as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }
      const newValue = {
        ...value,
        amount: inputValue,
      };
      onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
    }

    if (type === "Amount" || type === "AmountPercentage") {
      if (!isValidNumber(inputValue)) {
        const nativeEvent = new Event("onBlur", {
          bubbles: true,
          cancelable: true,
        });
        Object.defineProperty(nativeEvent, "target", {
          value: { value: inputValue },
          writable: false,
        });

        onBlur?.(nativeEvent as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }

      // Avoid transformation if the value is not a number
      if (isNotSafeValue(inputValue)) {
        const newValue = inputValue;
        const nativeEvent = new Event("onBlur", {
          bubbles: true,
          cancelable: true,
        });
        Object.defineProperty(nativeEvent, "target", {
          value: { value: newValue },
          writable: false,
        });

        onBlur?.(nativeEvent as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }
      const formatValue = displayValueAmount(
        inputValue,
        precision,
        viewPrecision,
        trailingZeros,
      );
      const nativeEvent = new Event("onBlur", {
        bubbles: true,
        cancelable: true,
      });
      Object.defineProperty(nativeEvent, "target", {
        value: { value: formatValue },
        writable: false,
      });

      onBlur?.(nativeEvent as unknown as React.FocusEvent<HTMLInputElement>);
      return;
    }

    // Handle the blur event for AmountCurrencyUniversal
    if (type === "AmountCurrencyUniversal" && typeof value === "object") {
      if (
        isValidBigInt(inputValue) &&
        isValidNumberGreaterThanMaxSafeInteger(inputValue)
      ) {
        const formatValue = inputValue;
        const nativeEvent = new Event("onBlur", {
          bubbles: true,
          cancelable: true,
        });
        Object.defineProperty(nativeEvent, "target", {
          value: { value: formatValue },
          writable: false,
        });
        onBlur?.(nativeEvent as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }
      // Avoid transformation if the value is not a number
      if (!isValidNumber(inputValue)) {
        const newValue = {
          ...value,
          amount: inputValue,
        };
        const nativeEvent = new Event("onBlur", {
          bubbles: true,
          cancelable: true,
        });
        Object.defineProperty(nativeEvent, "target", {
          value: { value: newValue },
          writable: false,
        });
        onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }
      const formatValue =
        viewPrecision !== undefined
          ? trailingZeros
            ? parseFloat(inputValue).toFixed(viewPrecision)
            : parseFloat(
                parseFloat(inputValue).toFixed(viewPrecision),
              ).toString()
          : inputValue;
      if (isNotSafeValue(inputValue)) {
        const newValue = {
          ...value,
          amount: inputValue,
        };
        onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
        return;
      }
      const newValue = {
        ...value,
        amount: formatValue,
      };
      onBlur?.(newValue as unknown as React.FocusEvent<HTMLInputElement>);
    }
  };

  const handleIsInputFocused = () => {
    setInputFocused(true);
  };

  return {
    isPercent,
    isShowSelect,
    options,
    valueInput: displayValueAmountState,
    valueSelect,
    handleOnChangeInput,
    handleOnChangeSelect,
    handleBlur,
    isBigInt,
    handleIsInputFocused,
    inputFocused,
    isAmount,
  };
};
