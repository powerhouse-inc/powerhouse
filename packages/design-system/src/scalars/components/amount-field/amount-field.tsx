import {
  AmountInput,
  type AmountInputProps,
} from "../../../ui/components/data-entry/amount-input/amount-input.js";
import { withFieldValidation } from "../fragments/with-field-validation/with-field-validation.js";
import { type FieldErrorHandling } from "../types.js";
import { validateAmountCurrency } from "./amount-currency-validations.js";
import { validateAmount } from "./amount-field-validations.js";

type AmountFieldProps = AmountInputProps & FieldErrorHandling;

const AmountField = withFieldValidation<AmountFieldProps>(AmountInput, {
  validations: {
    _numericAmount: validateAmount,
    _validOptionCurrency: validateAmountCurrency,
  },
});

AmountField.displayName = "AmountField";

export { AmountField, type AmountFieldProps };
