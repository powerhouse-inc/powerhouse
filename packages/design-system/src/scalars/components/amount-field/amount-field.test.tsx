import { commonCryptoCurrencies } from "#scalars";
import { screen, waitFor } from "@testing-library/react";
import { renderWithForm } from "../../lib/testing.js";
import { AmountField } from "./amount-field.js";

describe("AmountField Component", () => {
  it("should match snapshot", () => {
    const { container } = renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render label when provided", () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
      />,
    );
    expect(screen.getByLabelText("Amount Label")).toBeInTheDocument();
  });

  it("should render error messages when provided", async () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        errors={["Error 1", "Error 2"]}
        validators={() => "Error 3"}
        units={commonCryptoCurrencies}
      />,
    );
    await waitFor(() => {
      expect(screen.getByText("Error 1")).toBeInTheDocument();
      expect(screen.getByText("Error 2")).toBeInTheDocument();
      expect(screen.getByText("Error 3")).toBeInTheDocument();
    });
  });

  it("should render the percentage sign if the type is percent", () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="AmountPercentage"
        value={345}
        step={0}
      />,
    );
    expect(screen.getByText("%")).toBeInTheDocument();
  });

  it("should disable the input when disabled prop is true", () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
        disabled
      />,
    );
    const input = screen.getByLabelText("Amount Label");

    expect(input).toBeDisabled();
  });

  it("should set the input as required when required prop is true", () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        required
        units={commonCryptoCurrencies}
      />,
    );
    expect(screen.getByRole("spinbutton")).toHaveAttribute("required");
  });

  it("should render the description when provided", () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          amount: 345,
        }}
        units={commonCryptoCurrencies}
        disabled
        description="This is a description"
      />,
    );
    expect(screen.getByText("This is a description")).toBeInTheDocument();
  });
});
