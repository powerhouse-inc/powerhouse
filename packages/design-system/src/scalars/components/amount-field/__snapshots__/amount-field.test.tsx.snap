// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AmountField Component > should match snapshot 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:"
      >
        Amount Label
      </label>
      <div
        class="relative flex items-center"
      >
        <input
          data-cast="AmountNumber"
          name="amount"
          type="hidden"
        />
        <div
          class="relative flex items-center"
        >
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="relative flex items-center"
            >
              <input
                aria-invalid="false"
                class="flex h-9 w-full text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-white disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 pr-8 border border-r-[0.5px] border-gray-300 focus:border-r-0 rounded-md"
                data-cast="Number"
                id=":r1:"
                inputmode="numeric"
                name=""
                role="spinbutton"
                type="text"
                value="345"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
